### PR TITLE
internal/router: return 404 on meta/any with no includes

### PR DESCRIPTION
**Pull request by rogpeppe**
When we requested metadata on a fully qualified id using meta/any
with no includes, it would succeed unconditionally.

A better fix would be to make meta/any with no includes return an
error, but we can't do that without compromising the existing GUI client.

As drive-by fixes, we make the bulk id handler work correctly when
there are some ids are not found, and the error message from
the any handler return the full name of the meta endpoint requested.

### CODE REVIEW

File: internal/router/router.go [hunk 1]
```diff
@@ -285,6 +288,20 @@ func (r *Router) serveMetaGet(id *charm.Reference, req *http.Request) (interface
...
+			exists, err := r.handlers.Exists(id)
+			if err != nil {
+				return nil, errgo.Notef(err, "cannot determine existence of %q", id)
+			}
+			if !exists {
+				return nil, errgo.WithCausef(nil, params.ErrNotFound, "")
+			}
+			return params.MetaAnyResponse{Id: id}, nil
+		}
 		meta, err := r.GetMetadata(id, req.Form["include"])
```
Comment (mhilton) [inline internal/router/router.go]: Why not use includes here now that it exists.

File: internal/router/router.go [hunk 2]
```diff
@@ -124,6 +124,9 @@ type Handlers struct {
 	// which may end in a trailing slash (/) to indicate that longer
 	// paths are allowed too.
 	Meta map[string]BulkIncludeHandler
+
+	// Exists should report whether an entity with the given id exists.
+	Exists func(id *charm.Reference) (bool, error)
```
Comment (mhilton) [inline internal/router/router.go]: This seems out of place to me, I suspect this function is stored here in order to make writing tests easier, but I don't see that it is part of routing requests themselves. This bit seems oddly specific.

**Comment by jrwren** [comment]
## Test 1

I tested with an invalid url of no id and no query parameters and received a 500 response, not a 404.

```
$ http :8080/v4/meta/any
HTTP/1.1 500 Internal Server Error
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 56
Content-Type: application/json
Date: Wed, 19 Nov 2014 16:42:59 GMT

{
    "Code": "",
    "Message": "no ids specified in meta request"
}
```

## Test 2

A valid id with no query parameters

```
$ http :8080/v4/precise/elasticsearch/meta/any
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 35
Content-Type: application/json
Date: Wed, 19 Nov 2014 16:44:22 GMT

{
    "Id": "cs:precise/elasticsearch-0"
}
```

## Test 3

A valid id with extra-info

```
$ http :8080/v4/precise/elasticsearch/meta/any?include=extra-info
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Type: application/json
Date: Wed, 19 Nov 2014 16:54:47 GMT
Transfer-Encoding: chunked

{"id":..., "Meta"...
```

## Test 4

Same as 3 but using /v4/meta/any?id=...

```
$ http :8080/v4/meta/any?include=extra-info\&id=precise/elasticsearch
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Type: application/json
Date: Wed, 19 Nov 2014 16:56:21 GMT
Transfer-Encoding: chunked

{
    "precise/elasticsearch": {
        "Id": "cs:precise/elasticsearch-0",
        "Meta": {
            "extra-info": {
```

## Test 5

Same as test 2 but using /v4/meta/any?id=...

```
$ http :8080/v4/meta/any?id=precise/elasticsearch
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 61
Content-Type: application/json
Date: Wed, 19 Nov 2014 16:57:09 GMT

{
    "precise/elasticsearch": {
        "Id": "cs:precise/elasticsearch-0"
    }
}
```

From the description of the PR & commit message, I expected from 404 response on Test 2 and 5 , but I did not get them.

**Comment by mhilton** [comment]
I can't shake the idea that this is out of place in some way. However I think it would be better if the exists function were an unexported member of Router like resolveURL is, that would seem a better match.

**Comment by jrwren** [comment]
# QA try #2 - Now that I understand what this change.

## Fully qualified non-existing returns 404 just like partially qualified non-existing

```
$ http :8080/v4/precise/elasticsearc/meta/any?include=extra-info
HTTP/1.1 404 Not Found
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 92
Content-Type: application/json
Date: Wed, 19 Nov 2014 17:09:33 GMT

{
    "Code": "not found",
    "Message": "no matching charm or bundle for \"cs:precise/elasticsearc\""
}

$ http :8080/v4/precise/elasticsearc-0/meta/any?include=extra-info
HTTP/1.1 404 Not Found
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 90
Content-Type: application/json
Date: Wed, 19 Nov 2014 17:09:39 GMT

{
    "Code": "not found",
    "Message": "no matching charm or bundle for cs:precise/elasticsearc-0"
}
```

Should the /v4/meta/any behave the same way if all the ids do not exist?

```
$ http :8080/v4/meta/any?id=precise/elasticsearc-1
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 2
Content-Type: application/json
Date: Wed, 19 Nov 2014 17:08:20 GMT

{}
$ http :8080/v4/meta/any?id=precise/elasticsearc
HTTP/1.1 200 OK
Access-Control-Allow-Headers: X-Requested-With
Access-Control-Allow-Origin: *
Content-Length: 2
Content-Type: application/json
Date: Wed, 19 Nov 2014 17:08:24 GMT

{}
```

:+1: if the /v4/meta/any?id=DOESNOTEXIST should be 200

**Comment by mhilton** [comment]
:+1: With a small change of location. No QA

File: internal/router/router.go (hunk 1 above)
Comment (rogpeppe) [inline internal/router/router.go]: done

File: internal/router/router.go (hunk 2 above)
Comment (rogpeppe) [inline internal/router/router.go]: done

**Comment by bac** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**